### ISSUE TITLE
Rethink our APIs for table and shard creation related UDFs

### ISSUE THREAD

**Issue by ozgune** [citusdata/citus#694]
We currently provide the following APIs to users who want to create distributed tables.
- `master_create_distributed_table(table_name, distribution_column, distribution_method)` where the `distribution_method` is `append` or `hash`.
- `master_create_worker_shards(table_name, shard_count, replication_factor)`

We'd like to improve on these APIs in the following ways:
1. Introduce the notion of table colocation to our APIs. We currently enable colocation in an indirect manner: if the user picks the same distribution column, shard count, and replication factor, the tables end up being co-located. Our APIs should enable table colocation in an explicit and intuitive way.
2. Introduce the notion of reference tables. We currently enable reference tables in an indirect manner: if the user specifies a shard count less than `citus.large_table_shard_count`, we assume during query planning that the table can be broadcasted. This introduces confusion and we should enable creating reference tables in an intuitive way.
3. Simplify the notion of `distribution_method`. Our users currently use `append` distribution to shard tables on timestamp. We could potentially call this distribution method "distribution on time" and require that the user picks a timestamp column. We could also potentially pick `hash` as the default method.
4. The user currently needs to call two UDFs to make their table distributed. One motivation for this was due to the asymmetry between `append` and `hash` distribution. A second motivation was that this made it easier to reason about our failure semantics. We should look to see if unifying these functions could make sense.
5. Provide sensible default for `shard_count`. The shard count setting is the one we receive the most number of questions on from early users. We should look to calculate a sensible value here if we can infer the cluster size and total number of cpu cores. This is easier to do on the Cloud. If we can't compute a good value for this, we might pick a good enough default value for new users.
6. Remove confusion around `replication_factor`. We default to a replication_factor of 2 in Citus. However, Citus Cloud uses streaming replication for its replication model and therefore expects this value to be 1. We should look to hide this complexity from the user by potentially defaulting to 1 or 2 based on the default replication model.

We don't need to make all of these improvements together. It's still worthwhile thinking about them holistically.

**Comment by ozgune** [comment]
We talked about improving our APIs in two email threads. I'm copy/pasting our first email thread below (re: Better create distributed table functions).

--
The current functions to create distributed tables are a bit confusing. To start the conversation, I'd like to make the following suggestions as an alternative:

<pre>
-- Create a replicated (broadcast) table
SELECT create_replicated_table('users');

-- Create a hash-distributed table with 16 shards, honouring citus.shard_replication_factor
SELECT create_hash_distributed_table('events', 'userid', 16);

-- Create append-distributed table
SELECT create_append_distributed_table('events', 'userid');

-- Create a table that is co-located with events (must be hash-distributed)
SELECT create_colocated_table('users', 'userid', 'events');
</pre>

Our join logic then becomes:

replicated + replicated = broadcast join (on any column, may be stale)
replicated + distributed = broadcast join (on any column, may be stale)
distributed + colocated = co-located join (on partition column only, up-to-date)
distributed + distributed = re-partition join  (on any column, up-to-date, heavy network traffic)

--
I'm very much in favor of dropping it down to one function for setup. 

--
This would be an awesome improvement!

But is it useful to specify the number of shards? Can we simplify by making the call 

SELECT create_hash_distributed_table('events', 'userid');

and automatically creating some large number like 1024 shards (or maybe it's a configuration parameters and all tables have the same #), then if a join matches a partition column with a partition column we know it's a co-located join?

--
>> -- Create a replicated (broadcast) table
SELECT create_replicated_table('users');

>> -- Create a hash-distributed table with 16 shards, honouring citus.shard_replication_factor
SELECT create_hash_distributed_table('events', 'userid', 16);

>> -- Create append-distributed table
SELECT create_append_distributed_table('events', 'userid');

Is there a model where we can specify shards? Or pre-create empty shards? 

Then, maybe a function like expand_shard_count('events', +100) (which would error out for hash-distributed tables for now).  Related to this, is there a way to make range-partitioned tables more of a first-class citizen? If we are re-thinking API's, it might be time to think about it.

--
https://github.com/postgrespro/pg_pathman has a pretty nice API for creating range-partitioned tables. We could so something similar for shards in distributed tables.

--
After our call in the morning, we also chatted some more about this at the office. Two important points that came up related to thinking through the user's experience more holistically.

First, we talked about the types of tables a user can create in Citus: (a) distributed -> sharded by hash, (b) distributed -> sharded by append, (c) distributed -> reference, (d) local -> intermediate or temporary, and (e) local Postgres tables. We currently don't differentiate communicate how/when one can use these tables. Also, we haven't quite mapped out these tables to different use-cases.

Second, we talked about what it would take to make broadcast and co-located joins successful. I'm noting a few of these notes below:
1. Our co-location strategy happens to work if the user specifies the same number of shards and nodes. If the user runs the shard rebalancer, bad things will happen.
2. If we made co-location explicit, INSERT INTO (SELECT ...) becomes easier to implement. Other features, such as foreign key constraints might also become easier.
3. Ideally, we wouldn't concern the user with the number of shards.
4. If we introduce reference tables, does the user then only insert through the master node?
5. For left joins between a reference and sharded table, our current hypothesis is that the join happens on the "primary key". Do we then want to guide the user into making the reference table a sharded (by hash) table as well?
An important (open) follow-up question was then on whether we'd like to simplify the user experience. As it stands, different features have different gotchas and we're doing little communicating each one's implication to the user.

--
>> SELECT create_hash_distributed_table('events', 'userid');

>> and automatically creating some large number like 1024 shards (or maybe it's a configuration parameters and all tables have the same #), then if a join matches a partition column with a partition column we know it's a co-located join?

We could definitely use a default value, though it should probably be something like 120 (which is nicely divisible). 

The trouble with hash-distribution is that you either run a query on one shard, or all shards; there's no middle-ground. If you need to make 1024 connections to run a parallel query, then on small clusters you can only run 0-1 parallel queries at a time unless you put pgbouncer in front of the workers. There is also some per-shard overhead, which is typically about 1ms, so 1024 would cause parallel queries to take >1s.

--
Since we are accumulating thoughts, I think what would be useful in the append distribution is saying that it would be used *only* for time columns and being able to specify a time-period (instead of shard count or shard size).

API could be something like: ``create_append_distributed_table(tablename, timestamp columnname, duration);`` and we would create one shard per "duration".

So, something like ``create_append_distributed_table('events', 'userid', 1 hour)`` would create an hourly shard.

This won't be only an API change but also will require implementation changes + some auto-creating of shards etc. (like proper "range" support).

If we have clear time based shard boundaries, then expiring data for particular days, moving to cstore at end of day, caching results for particular day etc. would also become easier without the user having to ensure non-overlapping shards etc. and having to deal with the create_empty_shard / append_table_to_shard APIs.

**Comment by ozgune** [comment]
I'm copy/pasting our second email thread on this issue (re: Thoughts - table types and user-facing APIs)

--
A few thoughts / unbaked ideas below, reflecting on our conversation yesterday morning. 

create_distributed_table (<table_name>, <table_type>, <distribution_column>, <optional: co-locate table>, other optional parameters).

Examples - forgive the sloppy syntax, just review the abstraction:
<pre>
(1) create_distributed_table ("myProducts", reference_table)
(2) create_distributed_table ("myEvents", distribute_by_time, "someDateField")
(3a) create_distributed_table ("myEvents", distribute_by_id_hash, "someIdField")
(3b) create_distributed_table ("myEvents", distribute_by_id_hash, "someIdField", "someOtherHashedTableName") --> co-locates with the other table.
(4) create_distributed_table("myEvents", distribute_and_partition, "someIdField1", "someDateField")
--> customer x use-case.
</pre>

We could just pick (3a) as a starting point on Citus-MX, and worry about all others later on. Or not :) don't know the implementation work, but just getting the conversation started on the interface here.

--
Adding a few thoughts to the below while I'm at it, some of it also inline:

(I) In my ideal world as a user, I'd want to see as little complexity/friction/unfamiliar things as possible when I get started. And I get started with creating a table -- e.g. not having to specify the shard count; or shard size for append  -- or ideally even distribution method. My database would take care of all this automatically with some smart defaults. I realize all the below combined is a lot of work to get right, but want to put it out there as an aspirational end-goal to get the User Interface/API conversation started.
   --> Once we have the end-goal in mind, then we lay out the individual pieces, and start with the highest impact one on that path.

(II) Not knowing the implementation paths across products here, I'll be happy if we have  some variant of only (2) and (3a) below for a starting point -- and perhaps only (3a) for MX. Then (1) where we have broadcast joins; or (3b) where we have colocated joins.

**Comment by ozgune** [comment]
I wanted to provide an update on this issue with our upcoming APIs for v6.0.

(1) We decided to provide our users with two simple APIs. We will still keep our existing APIs for backwards compatibility and our new APIs will call the existing ones behind the covers. (#694)

<pre>
CREATE TABLE users (...);
SELECT create_distributed_table('users', 'user_id');
</pre>

This API will create a distributed table with hash distribution, and create shards on worker nodes with default shard count (32) and shard replication factor (already in guc). We will read both of these values from GUC. (#802)

<pre>
CREATE TABLE region (...);
SELECT create_reference_table('region');
</pre>

This API will create a reference table that only has one shard. The API will pick a distribution column for the reference table (any column will do). Although not required, we could pick one of the primary key columns and if not, any column that has a base type.

(2) We discussed four approaches for implementing colocation. We're going to go with the fourth approach. (#695)

<pre>
CREATE TABLE users (...);
SELECT create_distributed_table('users', 'user_id');
=> no other tables exists, not colocated with anything

CREATE TABLE events (..);
SELECT create_distributed_table('events', 'user_id');
=> behaviour 1), i.e. colocates on type/shardcount/replication factor

CREATE TABLE events_otherdimension (..);
SELECT create_distributed_table(‘events_otherdimension’, 'user_id', colocate_with => ''); 
=> colocate with nothing
</pre>

(3) Other: (a) We have three minor open implementation questions; and I'll drop updates on them to respective issues. (b) I'm going to follow up on the final function names for the new APIs, particularly the name we pick for "reference" tables. (c) We also haven't yet discussed the question of a default value for ``shard_replication_factor`` (already exists in GUC). I'll follow up on this later this week.

(4) We also decided to break up these tasks and prioritize them in the following order:

1. create_distributed_table() infers and automatically colocates tables
2. We read defaults for shard count and shard replication factor from GUC
3. create_reference_table() creates small tables
4. create_distributed_table() overrides colocation with colocate_with option
5. We implement logic to upgrade existing users to the new colocation apis
